### DISCUSSION TITLE
How to check existence of portalled elements with testing libraries?

### DISCUSSION THREAD

**Discussion by seedy** [radix-ui/primitives#1130] [Help] [answered]
Hi, folks! Hope you're doing well :wave: 

I'm using [Testing-Library/React](https://testing-library.com/docs/react-testing-library/intro) to perform integration tests in my application.

I would need to validate some internal rules where some user permissions impact what's rendered inside a `DropdownMenuContent`.

I need to have it `portalled` for my UI.

I don't use the [portal utility](https://www.radix-ui.com/docs/primitives/utilities/portal) yet.

In my DOM, I have a `<div data-radix-portal >` node at the end of my `body`, where I can see the `DropdownMenuContent` mounts when opened.
Hence I would have expected to look for it in my tests to find any portalled element.

I noticed that the `<div data-radix-portal >` node is not rendered inside my testing environment, I was wondering if you had clues about the why?

<details>

<summary>Repro</summary>

Here's my simplest repro as of now:

```
import * as DropdownMenuPrimitive from '@radix-ui/react-dropdown-menu';
import {render} from '@testing-library/react'

const DropdownMenu = DropdownMenuPrimitive.Root;
const DropdownMenuTrigger = DropdownMenuPrimitive.Trigger;
const DropdownMenuContent = DropdownMenuPrimitive.Content;
const DropdownMenuGroup = DropdownMenuPrimitive.Group;
const DropdownMenuItem = DropdownMenuPrimitive.Item

  test.only('dropdown', async () => {
    const Component = () => (
      <DropdownMenu modal>
        <DropdownMenuTrigger asChild>
          <button data-testid={`button`}>
            ...
          </button>
        </DropdownMenuTrigger>
        <DropdownMenuContent align="end" data-testid={`menu`}>
          <DropdownMenuGroup>
            <DropdownMenuItem>View</DropdownMenuItem>
            <DropdownMenuItem >
              Delete
            </DropdownMenuItem>
          </DropdownMenuGroup>
        </DropdownMenuContent>
      </DropdownMenu>
    )

    const { debug, container } = render(<Component />, {
      container: document.body,
    });

    act(() => {
      const button = getByTestId('button');
      fireEvent.click(button);
    })
    debug(container)
  })
```

# Logs
```
     <body>
        <button
          aria-haspopup="menu"
          data-state="closed"
          data-testid="button"
          id="radix-id-3792579812-1"
          type="button"
        >
          ...
        </button>
      </body>
```

</details>

Also, what would be the best practice you'd recommend to properly manage portals inside tests?

# A few moments Later...

Portal's implementation [here](https://github.com/radix-ui/primitives/blob/main/packages/react/portal/src/Portal.tsx)

This specific [line](https://github.com/radix-ui/primitives/blob/main/packages/react/portal/src/Portal.tsx#L35) made me wonder whether my testing environment doesn't follow the expected structure :confused: 

- When I `jest.spyOn` `ReactDOM.createPortal`, I see no call to it.
- When I log `globalThis?.document?.body`, I see it is exactly equal to `document.body`

<details>

## Test

```
test('dropdown', async () => {
  const portalSpy = jest.spyOn(ReactDOM, 'createPortal');

  const Component = () => (
    <DropdownMenu modal>
      <DropdownMenuTrigger asChild>
        <button data-testid={`button`}>
          ...
        </button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" data-testid={`menu`}>
        <DropdownMenuGroup>
          <DropdownMenuItem>View</DropdownMenuItem>
          <DropdownMenuItem >
            Delete
          </DropdownMenuItem>
        </DropdownMenuGroup>
      </DropdownMenuContent>
    </DropdownMenu>
  )

  const { getByTestId } = render(<Component />);

  act(() => {
    const button = getByTestId('button');
    fireEvent.click(button);
  })

  console.log(globalThis?.document?.body);
  console.log('equals documentBody', globalThis?.document?.body === document.body)
  screen.debug();

  console.log(portalSpy.mock.calls.length)

  await waitFor(() => {
    expect(getByTestId('menu')).toBeInTheDocument()
  })
})
```

## Logs

```
console.log
    HTMLBodyElement {}

      at Object.<anonymous> (__tests__/portalled-dropdown.spec.tsx:40:11)

  console.log
    equals documentBody true

      at Object.<anonymous> (__tests__/portalled-dropdown.spec.tsx:41:11)

  console.log
    <body>
      <div>
        <button
          aria-haspopup="menu"
          data-state="closed"
          data-testid="button"
          id="radix-id-965528199-1"
          type="button"
        >
          ...
        </button>
      </div>
    </body>
```

</details>

**Comment by vorillaz** [comment]
Hey there, you can use the container option upon rendering the dropdown component 
https://testing-library.com/docs/react-testing-library/api/#container

```jsx
const {container} = render(<Dropdown {...props} />, {
  container: document.body,
});
```

**Reply by seedy** [reply]
@vorillaz I just updated my question with a simple code sample. Trying to debug my container shows no sign of my portalled dropdown :sweat_smile:

**Reply by vorillaz** [reply]
You should fire a click event onto the trigger in order to see the modal popping, otherwise the portal is empty.
Regarding your question about portals, I am always in favour of mocking portals as:
```tsx
beforeAll(() => {
  ReactDOM.createPortal = jest.fn((element) => {
    return element;
  });
});

afterAll(() => {
  ReactDOM.createPortal.mockClear();
});
```

**Reply by seedy** [reply]
@vorillaz good point on the `fireEvent`, I totally forgot :facepalm: 

I don't think your proposal about mocking `createPortal` solves my issue, though.

How would I check the content of my dropdown then ?


By the way, you should check the signature of `createPortal`, it expects 2 params:
```
function createPortal(children: React.ReactNode, container: Element, key?: string): React.ReactPortal
```

**Reply by seedy** [reply]
I just noticed that spying on `ReactDOM.createPortal` doesn't help here:

```

  const portalSpy = jest.spyOn(ReactDOM, 'createPortal');

  const { debug, container, getByTestId } = render(<Component />, {
    container: document.body,
  });

  act(() => {
    const button = getByTestId('button');
    fireEvent.click(button);
  })
  expect(portalSpy).toHaveBeenCalled()

```

`Received number of calls: 0`

**Reply by vorillaz** [reply]
So, that was a sample as a reference. In theory you should follow a strategy that suits your use case scenarios and the idea behind your testing strategy.
In a nutshell, a combination of Jest and react-testing-library runs on top of JSDom which mimics the browser's environment. Dropdowns and other UI elements are bound with UI interactions and they are tighten to specific behaviors: once you click outside of a dropdown the contents are removed from the DOM.

Jest on the other hand can run multiple tests in parallel, thus you may encourage multiple dropdowns mounted at the same time on the DOM as the suite is running.

Using mocking you can pass along the rendered components and validate your logic. 
https://github.com/facebook/react/issues/11565#issuecomment-380143358

**Reply by seedy** [reply]
@vorillaz 
My testing strategy is to check rendered elements inside my DOM. Sorry if that was not clear enough from the start.

I am mentionning the fact that I want to check the content of my portal.

I am using `@testing-library` to that end

As you can see from my initial question, my goal is to understand why Radix-UI's portal is not **rendered** inside my DOM and find a way to **render** it inside the said DOM.

**Comment by Eddiejjay** [comment]
Hey! Facing same problem, did you solve how to get access portal in tests ?

**Reply by seedy** [reply]
Nope :confused: 

I must confess I didn't prioritize this task a lot, recently.

I have some clues for workarounds, but that's about it for now...

### Dialog

The `Dialog` component has a `Portal` subcomponent, which might help forcing its parts to render wherever you wish during testing.

See https://www.radix-ui.com/docs/primitives/components/dialog#portal

I might try testing this!

### DropdownMenu

Avoid portalling.

I think the only way to test this component is not to use the portalled version, to do this, set `portalled={false}` on the `Content` subcomponent.

**Reply by seedy** [reply]
Better news: there seems to be ongoing work to reproduce the component model of `Dialog` and allow customizing the way the portalled components are rendered :tada: :
- https://github.com/radix-ui/primitives/issues/1306
- https://github.com/radix-ui/primitives/issues/1305
- https://github.com/radix-ui/primitives/issues/1304
- https://github.com/radix-ui/primitives/issues/1303

**Reply by vnugent** [reply]
I'm facing similar issue.  @seedy did you find a solution?

I can find and click on the trigger button in the test.  However, I don't see `DropdownMenuContent` being rendered under `<body>`.

**Reply by benoitgrelard** [reply]
Should it matter to your tests that it's portalled? Or should you check for menu items and click them? I would argue the former is just an implementation detail no?

**Reply by vnugent** [reply]
@benoitgrelard the latter. I'm in agreement with you that it's an internal implementation detail.  I don't want to interfere with the default behavior of [Dropdown primitive](https://www.radix-ui.com/docs/primitives/components/dropdown-menu) which is rendering the content to body.  

The issue here is I can't find the dropdown content anywhere.  

```javascript
  // ... setup omitted

  test('Popup menu items with permission to delete', async () => {
    const user = userEvent.setup()
    const { screen} = render(
      <PopupTagList
        list={TAG_DATA}
        imageInfo={IMAGE_DATA}
      />
    )
    expect(AddTagMock).toHaveBeenCalledTimes(1)

    await user.click(screen.getByRole('button')) // content should be rendered to body

    screen.debug()
 })
```

Console

```javascript
   <body> // <--- Radix Dropdown should be a child of body
      <div>
        <div aria-label="tag popup">
          <button>Menu trigger button</button>
          <div>Add tag mock</div>
       </div>  
    </div>
</body>
```

**Reply by JanderSilv** [reply]
Hey @vnugent , did you find out?

**Reply by djalmaaraujo** [reply]
@benoitgrelard Any ideas? I am also facing this issue. The dropdown content is nowhere. I tried a lot of things, and it simply does not render.

**Reply by josephadd** [reply]
Same with me now. Any has got any solution? @djalmaaraujo @vnugent @vorillaz @benoitgrelard

**Reply by hallucinogenizer** [reply]
Any luck anyone?

**Reply by wa1id** [reply]
This works for me, my `DropdownMenu` component is just abstracting the primitive parts into my own component:

```typescript
import { render, screen } from '@testing-library/react'
import { DropdownMenu, DropdownMenuContent, DropdownMenuItem, DropdownMenuTrigger } from './DropdownMenu'
import userEvent from '@testing-library/user-event'

describe('DropdownMenu', () => {
  it('renders the trigger', () => {
    render(
      <DropdownMenu>
        <DropdownMenuTrigger>Open Menu</DropdownMenuTrigger>
        <DropdownMenuContent>
          <DropdownMenuItem>Item 1</DropdownMenuItem>
          <DropdownMenuItem>Item 2</DropdownMenuItem>
        </DropdownMenuContent>
      </DropdownMenu>,
    )

    expect(screen.getByText('Open Menu')).toBeInTheDocument()
  })

  it('renders the items when trigger is clicked', async () => {
    const user = userEvent.setup()

    render(
      <DropdownMenu>
        <DropdownMenuTrigger>Open Menu</DropdownMenuTrigger>
        <DropdownMenuContent>
          <DropdownMenuItem>Item 1</DropdownMenuItem>
          <DropdownMenuItem>Item 2</DropdownMenuItem>
        </DropdownMenuContent>
      </DropdownMenu>,
    )

    await user.click(screen.getByText('Open Menu'))
    expect(screen.getByText('Item 1')).toBeInTheDocument()
    expect(screen.getByText('Item 2')).toBeInTheDocument()
  })
})
```

**Comment by imshubhamsingh** [comment, marked as answer]
Mocking the portal worked for me for Dialog and dropdown. We just wanted to check the functionality of the component as to what all get renders inside portal. Below is the mock:

```tsx
import { forwardRef } from "react";

import { PortalProps } from "@radix-ui/react-portal";

jest.mock("@radix-ui/react-portal", () => ({
  ...jest.requireActual("@radix-ui/react-portal"),
  Portal: forwardRef<HTMLDivElement, PortalProps>(function PortalMock(props, ref) {
    return <>{props.children}</>;
  }),
}));
```
